### PR TITLE
proposal: expose list of backed up resources in backup details

### PR DESCRIPTION
**Pull request by prydonius**
Signed-off-by: Adnan Abdulhussein <aadnan@vmware.com>

ref #396

### CODE REVIEW

File: design/backup-resource-list.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+configmaps:
+- default/database
+```
+
+The filename for this metadata will be `<backup name>-resource-list.json`.
+The top-level key is the string form of the `schema.GroupResource` type that we currently keep track of in the backup controller code path.
+
+### Changes in Backup controller
+
+The Backupper currently initialises a map to track the `backedUpItems` (https://github.com/heptio/velero/blob/1594bdc8d0132f548e18ffcc1db8c4cd2b042726/pkg/backup/backup.go#L269), this is passed down through GroupBackupper, ResourceBackupper and ItemBackupper. Moving the map initialisation to the BackupController will provide access to the resulting list there after a successful backup.
```
Comment (skriss) [inline design/backup-resource-list.md]: We're already a little bit inconsistent with how we return information from `pkg/backup` to the backup controller, but I think my preferred approach here would be to add a field to the `pkg/backup.Request` struct to store backed-up items. This can get populated by `pkg/backup` during execution of the `Backup(...)` method, and then the backup controller can read it when it goes to persist.  This is consistent with how volume snapshot metadata is passed currently.  

Alternately, we could define a new struct to capture returned info about the backup that would include volume snapshots, restic backup data (WIP by carlisia), and this info.

WDYT?

File: design/backup-resource-list.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+This command will need to be updated to fetch the metadata from the bucket using the `Stream` method used in other commands.
+The file will be read in memory and displayed in the output of the command.
+Depending on the format the metadata is stored in, it may need processing to print in a more human-readable format.
+If we choose to store the metadata in YAML, it can likely be directly printed out.
+
+If the metadata file does not exist, this is an older backup and we cannot display the list of resources that were backed up.
+
+## Open Questions
+
+- Do we want to show the group version in the metadata (i.e. do we want users to see `apps/v1/Deployment` instead of `deployments.apps`)? If so, we wouldn't be able to use the existing list of backedUpItems list, or we'd need to change it to track schema.GroupVersionKinds instead.
```
Comment (skriss) [inline design/backup-resource-list.md]: I think this would be nice to have -- particularly for the use case where a user is restoring an older backup into a new cluster, and needs to ensure that the backed-up API versions exist in their new cluster before restoring. What do others think?

File: design/backup-resource-list.md (hunk 1 above)
Comment (prydonius) [reply design/backup-resource-list.md]: Ah, I thought the backup.Request was the CR, but I see now that it's a different internal struct. I agree that it makes sense to use this or another struct to pass down the result.

Defining a new struct (i.e. backup.Response) will help make it more clear what it is used for, but using the Request struct isn't too terrible and won't require as many changes. What do others think?

File: design/backup-resource-list.md (hunk 2 above)
Comment (prydonius) [reply design/backup-resource-list.md]: That's a very good point, do you think we can change backedUpItems to track with versions? I wasn't sure if it relies on unambiguous versions to make sure there are no duplicates?

Comment (skriss) [reply design/backup-resource-list.md]: Velero only backs up the server's "preferred version" of each API group, so in theory we should only have one version per resource. We could update the `itemKey` struct to use a GVR rather than just a GR, which should work OK (need to figure out details though).

File: design/backup-resource-list.md (hunk 1 above)
Comment (skriss) [reply design/backup-resource-list.md]: 👍using the `backup.Request` WFM.

File: design/backup-resource-list.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+- default/database
+- default/wordpress
+services:
+- default/database
+- default/wordpress
+secrets:
+- default/database-root-password
+- default/database-user-password
+configmaps:
+- default/database
```
Comment (nrb) [inline design/backup-resource-list.md]: An example of cluster-scoped resources would be helpful.

I assume it would look something like:

```yaml
persistentvolumes:
- my-pv
```

File: design/backup-resource-list.md (hunk 2 above)
Comment (nrb) [reply design/backup-resource-list.md]: Agreed that writing the version of the backed up item would be nice, especially since we should only have one per item.

File: design/backup-resource-list.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+### Fetch backup contents archive and walkthrough to list contents
+
+Instead of recording new metadata about what resources have been backed up, we could simply download the backup contents archive and walkthrough it to list the contents everytime `velero backup describe <name> --details` is run.
+
+The advantage of this approach is that we don't need to change any backup procedures as we already have this content, and we will also be able to list resources for older backups.
+Additionally, if we wanted to expose more information about the backed up resources, we can do so without having to update what we store in the metadata.
+
+The disadvantages are:
+- downloading the whole backup archive will be larger than just downloading a smaller file with metadata
+- doesn't solve the usecase of wanting to see the list of backed up resources in the bucket outside of Velero tooling
```
Comment (nrb) [inline design/backup-resource-list.md]: I'm not 100% sure this is a design goal for Velero as a whole, especially if/when we add encryption. That said, we've also never purposefully worked to obfuscate what we read from the bucket, so using tar listing is going against the status quo.

File: design/backup-resource-list.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+- Include a list of backed up resources as metadata in the bucket
+- Enable users to get a view of what resources are included in a backup using the Velero CLI
+
+## Non Goals
+
+- Expose the full manifests of the backed up resources
+
+## Background
+
+As reported in #396, the information reported in a `velero backup describe <name> --details` command is fairly limited, and does not easily describe what resources a backup contains.
```
Comment (nrb) [inline design/backup-resource-list.md]: May be worth adding the actual link to the issue at `#396`.

Comment (prydonius) [reply design/backup-resource-list.md]: huh, I thought it would be added by GitHub directly, will do!

File: design/backup-resource-list.md (hunk 4 above)
Comment (prydonius) [reply design/backup-resource-list.md]: I can't remember where I got this from, but I agree this is not exactly a usecase we're trying to solve for. Let me reword this to reduce the emphasis on this point.

**Comment by prydonius** [comment]
Updated to mention changing backedUpItems resource key to using schema.GroupVersionKind instead.

**Comment by prydonius** [comment]
@skriss @nrb PTAL!

File: design/backup-resource-list.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+The filename for this metadata will be `<backup name>-resource-list.json`.
+The top-level key is the string form of the `schema.GroupResource` type that we currently keep track of in the backup controller code path.
+
+### Changes in Backup controller
+
+The Backupper currently initialises a map to track the `backedUpItems` (https://github.com/heptio/velero/blob/1594bdc8d0132f548e18ffcc1db8c4cd2b042726/pkg/backup/backup.go#L269), this is passed down through GroupBackupper, ResourceBackupper and ItemBackupper where ItemBackupper records each backed up item.
+This property will be moved to the [Backup request struct](https://github.com/heptio/velero/blob/16910a6215cbd8f0bde385dba9879629ebcbcc28/pkg/backup/request.go#L11), allowing the BackupController to access it after a successful backup.
+
+`backedUpItems` currently uses the `schema.GroupResource` as a key for the resource.
```
Comment (nrb) [inline design/backup-resource-list.md]: Is there a function that allows for easy parsing of this format back into a GVR?

Comment (prydonius) [reply design/backup-resource-list.md]: Do you mean converting from `schema.GroupResource` to `schema.GroupVersionKind` or the other way around? We can't get the GVK from the GroupResource, but we can get it from the `runtime.Unstructured` obj. We don't really need the GroupResource, AFAICT we're only using it as a key in backedUpItems and not using it for anything else after.

Comment (nrb) [reply design/backup-resource-list.md]: Sorry, should have commented lower down. I meant the `{group}/{version}/{kind}` string, and I meant GVK, not GVR. I suppose writing a deserialization function would be simple enough.

It may not actually be all that important, as we could just print the key out verbatim when displaying details.

Comment (prydonius) [reply design/backup-resource-list.md]: Ah I see what you mean now, I was just thinking of printing out the key as-is, but we wouldn't be able to directly parse that string (since the GVK.String() method has a slightly different format), but (without looking at the schema pkg) I believe we can parse the `schema.GroupVersion` using `{group}/{version}` and then convert that to GVK using the remaining part.

File: design/backup-resource-list.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+In order to record the API group, version and kind for the resource, this key will be constructed from the object's `schema.GroupVersionKind` in the format `{group}/{version}/{kind}` (e.g. `apps/v1/Deployment`).
+
+The `backedUpItems` map is currently a flat structure and will need to be converted to the nested structure above, grouped by `schema.GroupVersionKind` string.
```
Comment (skriss) [inline design/backup-resource-list.md]: You may want to consider decoupling the way this data is stored internally in `pkg/backup` from how it gets written to file for display purposes. Having a flat map makes lookup very efficient during backup, whereas having something like a `map[schema.GroupVersionKind][]string` (which I think is what you're proposing), while closer to what you want for view purposes, would result in less efficient lookups during backup. Not a big deal for small clusters, but it could add some overhead in larger ones. I'm fine sorting this out as you move to implementation.

File: design/backup-resource-list.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+The filename for this metadata will be `<backup name>-resource-list.json`.
+The top-level key is the string form of the `schema.GroupResource` type that we currently keep track of in the backup controller code path.
+
+### Changes in Backup controller
+
+The Backupper currently initialises a map to track the `backedUpItems` (https://github.com/heptio/velero/blob/1594bdc8d0132f548e18ffcc1db8c4cd2b042726/pkg/backup/backup.go#L269), this is passed down through GroupBackupper, ResourceBackupper and ItemBackupper where ItemBackupper records each backed up item.
+This property will be moved to the [Backup request struct](https://github.com/heptio/velero/blob/16910a6215cbd8f0bde385dba9879629ebcbcc28/pkg/backup/request.go#L11), allowing the BackupController to access it after a successful backup.
+
+`backedUpItems` currently uses the `schema.GroupResource` as a key for the resource.
+In order to record the API group, version and kind for the resource, this key will be constructed from the object's `schema.GroupVersionKind` in the format `{group}/{version}/{kind}` (e.g. `apps/v1/Deployment`).
```
Comment (skriss) [inline design/backup-resource-list.md]: Is there a specific reason to use Kind rather than Resource? I don't think it makes much of a difference from a UX perspective, but the directories in the tarball do use "resource.group" as the name format so it will look slightly different (capitalization and pluralization generally).

File: design/backup-resource-list.md (hunk 7 above)
Comment (prydonius) [reply design/backup-resource-list.md]: You're right, I think I was envisioning that we would do this conversion when creating the metadata file (i.e. in persistBackup), so I think internally in pkg/backup it would keep the flat structure - I'll update this to make that more clear.

File: design/backup-resource-list.md (hunk 8 above)
Comment (prydonius) [reply design/backup-resource-list.md]: I guess the difference will be "apps/v1/deployments" vs "apps/v1/Deployment"? No specific reason, but Kind might be easier to read especially for resources with multiple words ("persistentvolume" vs "PersistentVolume").

File: design/backup-resource-list.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+v1/Secret:
+- default/database-root-password
+- default/database-user-password
+v1/ConfigMap:
+- default/database
+v1/PersistentVolume:
+- my-pv
+```
+
+The filename for this metadata will be `<backup name>-resource-list.json`.
```
[the remark this answers is not in the document]
Comment (nrb) [reply design/backup-resource-list.md]: Should we also store this in the `metadata` dir within the backup structure?

**Comment by nrb** [comment]
I think I'm :+1: on this in general. Once this is in I can see a few more feature requests being generated, such as being able to filter the items returned by this, but that can be addressed later.

File: design/backup-resource-list.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (skriss) [reply design/backup-resource-list.md]: It should just go within the backup's directory (`backups/<backup-name>/`)

File: design/backup-resource-list.md (hunk 7 above)
Comment (skriss) [reply design/backup-resource-list.md]: Ah, I see - SGTM

File: design/backup-resource-list.md (hunk 8 above)
Comment (skriss) [reply design/backup-resource-list.md]: OK, seems fine

File: design/backup-resource-list.md (hunk 7 above)
Comment (prydonius) [reply design/backup-resource-list.md]: @skriss reworded a bit to make this more clear

**Comment by skriss** [comment]
LGTM - let's update to "approved" and merge!
